### ISSUE TITLE
Can avifJPEGRead() set cinfo.out_color_space to JCS_YCbCr to avoid the conversion to/from RGB?

### ISSUE THREAD

**Issue by wantehchang** [AOMediaCodec/libavif#372]
Joe,

When I run avifenc in the debugger and dump the contents of the 'cinfo' local variable in avifJPEGRead(), I see the following values:

    jpeg_color_space = JCS_YCbCr, out_color_space = JCS_RGB

In jpeglib.h, these two fields of struct jpeg_decompress_struct are documented as follows:

```
  J_COLOR_SPACE jpeg_color_space; /* colorspace of JPEG image */

  /* Decompression processing parameters --- these fields must be set before
   * calling jpeg_start_decompress().  Note that jpeg_read_header() initializes
   * them to default values.
   */

  J_COLOR_SPACE out_color_space; /* colorspace for output */
```

That comment seems to suggest we could set cinfo.out_color_space to JCS_YCbCr between our jpeg_read_header() and jpeg_start_decompress() calls to avoid the conversion to/from RGB. Do you know if this will work?

I found this code in the third_party/blink/renderer/platform/image-decoders/jpeg/jpeg_image_decoder.cc file in the Chrome source tree:

https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/platform/image-decoders/jpeg/jpeg_image_decoder.cc;l=763?q=%22out_color_space%20%3D%20JCS_YCbCr%22&ss=chromium

```
      case JPEG_START_DECOMPRESS:
        if (decoding_mode == JPEGImageDecoder::DecodingMode::kDecodeToYuv) {
          DCHECK(decoder_->CanDecodeToYUV());
          DCHECK(decoder_->HasImagePlanes());
          info_.out_color_space = JCS_YCbCr;
          info_.raw_data_out = TRUE;
          uv_size_ = ComputeYUVSize(&info_, 1);
          // U size and V size have to be the same if we got here
          DCHECK_EQ(uv_size_, ComputeYUVSize(&info_, 2));
        }
```

Thank you.

**Comment by joedrago** [comment]
I have never tried to receive YUV from libjpeg before, but assuming it can emit YUV420 (or YUV444?), it should play nicely with the current image loader architecture in `avifenc`. The conversion from RGB to YUV already happens inside of each image decoder. 

I think the right play here (assuming it can be done) would perhaps be to auto-detect that the current CICP values from the commandline/defaults were *compatible* with JPEG's internal format (CICP `[2/1],[2/13],[2/5/6]`?), and if so, harvest the YUV directly. If the CICP chosen on the commandline wasn't compatible, we'd just do the path we're already doing.

**Comment by joedrago** [comment]
I hacked around a bit this morning at this just to test the viability, and it definitely can work, although there a lot of open questions/decisions:

* Would we make this the default, or require that the enduser requests this path?
* What cmdline restrictions would we have to impose? I think the CICP would have to match, and (at first) the bit depth. 
* It seems like it unpacks to YUV444 no matter what, unless I'm missing something. We'd need to know if it was internally YUV420 and honor that (or support either path depending on commandline?). 

I've attached a diff for the hacky code I was playing around with it. It currently chooses to go directly to YUV if YUV444 is requested, but this is just for local testing and not actually what we'd want to do in the long term.

[jpegyuv.txt](https://github.com/AOMediaCodec/libavif/files/5403816/jpegyuv.txt)

**Comment by tongyuantongyu** [comment]
Hi, I tried and figured out how to do this. The implementation is [here](https://github.com/tongyuantongyu/libavif/blob/raw_input_from_jpeg/apps/shared/avifjpeg.c). I'm creating a new function `avifJPEGReadPreserve` as the logic differs a lot.

First of all `jpeg_read_raw_data` should be used instead of  `jpeg_write_scanlines` if we want raw data. Then we should inspect each components (planes) to see if its chroma downsample mode is compatible with AVIF. ([JFIF](http://www.w3.org/Graphics/JPEG/jfif3.pdf) supports many more modes, includes the weird one at page 5)

JPEG can store pixels in RGB, and even allow subsampling under RGB. Examples can be generated using cjpeg:
```
cjpeg -rgb -sample 1x1,2x2,1x1 -outfile foo.jpg foo.png
```
This will produce an RGB JPEG file with green channel has 2x width and height than red and blue. Most program can handle it normally, and it is exactly what a 8bit, full range, YUV420 subsample, identity MC AVIF image means. But if we ask avifenc to do the same thing (`avifenc -r f -d 8 --yuv 420 --nclx 1/13/0 foo.png foo.avif`), it will complains and produce an image with wrong MC. If I force the MC to be 0, then no program can decode it.

Besides that, libjpeg can work on 8,9,10,11,12 bit samples ([source](https://github.com/joedrago/libjpeg/blob/master/jmorecfg.h#L33)), and libjpeg-turbo supports 8 and 12bit mode. This is not widely supported as libjpeg force you to decide only one and change it requires recompile (libavif is great here).

So above is what I've explored. The implementation should work fine with libjpeg version 6-9, with `BITS_IN_JSAMPLE` configured as 8/10/12. It will try to import samples and information from jpeg file without conversion if that's possible. I'm not sure where to generate a warning(or error) if libjpeg is configured in 9/11bit mode, as we can do literally nothing with it.

As for when to use it, I'm thinking adding a flag to let user explicitly ask for it, and then ignore all flag about depth, range or CICP (like what we do on y4m inputs). But I'm not sure how this will influence other parts like creating animated avif, and should we make this flag effective on other formats as well.
